### PR TITLE
Declunkify focal point picker dragging

### PR DESCRIPTION
**Pull request by stokesman**
This PR stemmed from realizing that maybe it wasn't just me and my modest computer experiencing clunky dragging in the focal point picker.
https://github.com/WordPress/gutenberg/issues/28487#issuecomment-768343307

Setting out to improve that, it didn't take much tinkering to figure out that the bulk of the poor performance comes from setting block attributes during the drag operation. It's super intensive because it's driven by `onMouseMove` and causes quite the render cascade.

While I'd hoped to find the solution with changes only to `FocalPointPicker`, the consuming blocks need some changes as well. The Cover and Media & Text blocks have been updated to set their attributes only at the end of a drag operation and to preview the focal point with an imperative (render-free) approach while dragging.

A complication ensued in updating the Media & Text block. To create an imperative routine for focal point preview, a ref to a `ResizableBox` was needed, and being that it was also wrapped by `withNotices`, both components needed updates to support ref forwarding. To support that in `withNotices` it seemed best to refactor it to a function component. Touching that component looks fraught with peril since I see no unit tests for it, but I'm hoping there may be some E2E coverage.

UPDATE:  Unit tests have been added for `withNotices`. A few new unit tests have also been added for `FocalPointPicker`.

## How has this been tested?
Making posts with Cover and Media & Text blocks and dragging their focal points like mad.

## Screen recordings (as specific to my machine as they may be)
**Before:**

https://user-images.githubusercontent.com/9000376/106662181-ed6f0180-6556-11eb-8361-eeef1630d399.mp4

**After**

https://user-images.githubusercontent.com/9000376/106662341-1ee7cd00-6557-11eb-9144-4308d91960cb.mp4

## Types of changes
* Enhancement
* Breaking change: `FocalPointPicker` calls `onChange` only at the end of a drag operation, whereas before it was called for each `mousemove` event. So it still works, just not as rapidly as it used to. (I wasn't totally sure it qualifies as breaking but it's how I've classified it in the changelog).
* New feature: `onDrag` added as an event for `FocalPointPicker`. This is now what's called each `mousemove`
* Fix: Focus of the `FocalPointPicker` draggable area and adjustment with arrow keys. (This was added #22531 but was no longer working) 
* Fix: `FocalPointPicker` fires `onDragStart` and `onDragEnd` (they were in the code already but did not actually fire).
* New feature:`ResizeableBox` is now a `forwardRef`*
* New feature: `withNotices` returns a `forwardRef` if passed a `forwardRef`*

*React docs suggest to treat changing a component to a `forwardRef` as breaking, but I believe that's for a context of changing a class component with an ad-hoc prop to support ref forwarding. In the cases of `ResizableBox` and `withNotices` they had no support for ref forwarding before this otherwise have no changes. So it seems only an unbreaking added feature.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Comment by jasmussen** [comment]
Wow, this is a night and day improvement. You keep making amazing PRs to the project, thank you so much! Big fan of this work — I'll trust the developers to help review the details, but in testing this, the experience is vastly improved.

File: packages/block-library/src/cover/edit.js [hunk 1]
```diff
@@ -355,6 +358,13 @@ function CoverEdit( {
 		isVideoBackground ||
 		( isImageBackground && ( ! hasParallax || isRepeated ) );
 
+	const imperativeFocalPointPreview = ( value ) => {
+		const [ styleOfRef, property ] = isDarkElement.current
+			? [ isDarkElement.current.style, 'objectPosition' ]
+			: [ blockProps.ref.current.style, 'backgroundPosition' ];
+		styleOfRef[ property ] = mediaPosition( value );
```
Comment (ItsJonQ) [inline packages/block-library/src/cover/edit.js:365]: This ternary + destructuring is really cool! Will add this little pattern to my mental list cool JS patterns :)

File: packages/components/src/resizable-box/index.js [hunk 2]
```diff
@@ -102,4 +111,4 @@ function ResizableBox( {
 	);
 }
 
-export default ResizableBox;
+export default forwardRef( ResizableBox );
```
Comment (ItsJonQ) [inline packages/components/src/resizable-box/index.js:114]: IMO, forwarding refs should be standard for components from @wordpress/components.
I'm happy to see more of these :)

**Review by ItsJonQ — APPROVED** [review]
🚀 from me! This is great @stokesman  :).

I think offering an `onDrag` as an "in-between" of start/stop (`onChange`) works well.

Like you've demonstrated, it provides the flexibility to handle updates/rendering in alternate ways. In this case, manipulating the DOM directly, by-passing React's render cycle (which is always faster).

The trade-off is, we'd need to replicate this setup every time - which I believe is uncommon, but may happen.

It's a 👍 from me.

---

(Below are my thoughts on Gutenberg / UI / performance. Not directly related to this PR :D)

I feel like this touches upon a deeper problem - one that I've encountered often in my work exploring improving Design Tools for Gutenberg. That is... how do we effectively add rich interactions without degrading performance? @stokesman from what I've observed of your PRs, this is probably something you're mindful of as well.

Rich interactions often involve actions (usually drag) updating many times a second. These updates need to move up to the data store and back down, and ripple React's render cycles. From my research, the best pattern to resolve these issues (at scale) is to liberally use (data) selectors + (component) memoization to control re-renders. The trick is to establish the necessary patterns (perhaps helper functions?) to reduce the verbosity of this "rigging" work.

**Comment by stokesman** [comment]
Thanks for the quick test drive @jasmussen and review @ItsJonQ!

I've pushed a bunch of mostly small commits to patch up a few things, add tests, and update docs + changelog. With that, I'll take this out of draft status.

**Comment by ItsJonQ** [comment]
That's great @stokesman ! I'm 👍 for this to merge if other folks are

**Review by aristath — APPROVED** [review]
LGTM :+1:

**Comment by dcalhoun** [comment]
@stokesman @ItsJonQ @aristath these changes appear to have introduced a breakage for mobile Gutenberg in the form of wordpress-mobile/gutenberg-mobile#3149. Specifically, it appears it may be related to the changes to the `withNotice` HOC.

This breakage is causing an app crash and is blocking the next release of the mobile apps this week, which is a time-sensitive matter. 😞 Would you be able to help us identify a fix or evaluate the possibility of reverting these changes until a fix is identified please? Thank you!

**Comment by dcalhoun** [comment]
To provide an update on https://github.com/WordPress/gutenberg/pull/28676#issuecomment-779460409, I believe I have identified a fix and am awaiting review in #29013.

**Pull request merged**